test/catalog/catalog_test.cpp

```diff
@@ -502,4 +502,64 @@ TEST_F(CatalogTests, GetIndexObjectsTest) {
+ */
+// NOLINTNEXTLINE
+TEST_F(CatalogTests, DDLLockTest) {
+  // txn2 is used to verify that older txns can't acquire the lock
+  auto *txn2 = txn_manager_->BeginTransaction();
+
+  // txn0 is used to verify conflict scenarios and that commits release the lock
+  auto *txn0 = txn_manager_->BeginTransaction();
+  auto accessor0 = catalog_->GetAccessor(txn0, db_);
+  EXPECT_NE(accessor0, nullptr);
+  auto ns_oid = accessor0->CreateNamespace("test_namespace0");  // Should succeed as txn0 acquires the lock
+  EXPECT_NE(ns_oid, catalog::INVALID_NAMESPACE_OID);
+  VerifyCatalogTables(*accessor0);  // Check visibility to me
+
+  ns_oid = accessor0->CreateNamespace("test_namespace4");  // Should work since txn0 already holds lock
+  EXPECT_NE(ns_oid, catalog::INVALID_NAMESPACE_OID);
+  VerifyCatalogTables(*accessor0);  // Check visibility to me
+
+  // txn1 is used to verify that concurrent txns cannot acquire lock, even if a newer timestamp
+  auto *txn1 = txn_manager_->BeginTransaction();
+  auto accessor1 = catalog_->GetAccessor(txn1, db_);
+  ns_oid = accessor1->CreateNamespace("test_namespace1");
+  EXPECT_EQ(ns_oid, catalog::INVALID_NAMESPACE_OID);  // Should fail to get the lock because txn0 holds it (txn1 > txn0)
+  txn_manager_->Abort(txn1);
+
+  auto accessor2 = catalog_->GetAccessor(txn2, db_);
+  ns_oid = accessor2->CreateNamespace("test_namespace2");
+  EXPECT_EQ(ns_oid, catalog::INVALID_NAMESPACE_OID);  // Should fail to get the lock because txn0 holds it (txn2 < txn0)
+
+  txn_manager_->Commit(txn0, transaction::TransactionUtil::EmptyCallback, nullptr);  // txn0 releases the lock
+
+  ns_oid = accessor2->CreateNamespace("test_namespace2");
+  EXPECT_EQ(ns_oid,
+            catalog::INVALID_NAMESPACE_OID);  // Should fail to get the lock because txn0 committed (txn2 < txn0)
+
+  txn_manager_->Abort(txn2);
+
+  // txn4 is used to verify that older txns can acquire the lock after an abort
+  auto *txn4 = txn_manager_->BeginTransaction();
+
+  // txn3 is used to verify that aborts release the lock
+  auto *txn3 = txn_manager_->BeginTransaction();
+  auto accessor3 = catalog_->GetAccessor(txn3, db_);
+  ns_oid = accessor3->CreateNamespace("test_namespace3");
+  EXPECT_NE(ns_oid, catalog::INVALID_NAMESPACE_OID);  // Should succeed as txn3 acquires the lock
+  txn_manager_->Abort(txn3);                          // txn3 releases the lock
+
+  auto accessor4 = catalog_->GetAccessor(txn4, db_);
+  EXPECT_NE(accessor4, nullptr);
+  VerifyCatalogTables(*accessor4);  // Check visibility to me
+  ns_oid = accessor4->GetNamespaceOid(
+      "test_namespace0");  // Should succeed as txn4 acquires the lock (txn4 < txn3) but txn3 aborted
+  EXPECT_TRUE(accessor4->DropNamespace(ns_oid));
+  ns_oid = accessor4->GetNamespaceOid("test_namespace0");
+  EXPECT_EQ(ns_oid, catalog::INVALID_NAMESPACE_OID);
+  txn_manager_->Commit(txn4, transaction::TransactionUtil::EmptyCallback, nullptr);  // txn4 releases the lock
+}
```

Comment:
Missing sequence:
```
T1              T2
--------------------------------
BEGIN
                BEGIN
TryLock
COMMIT
                TryLock (fails)
```

Still need to test because currently tested via an undefined code path.
```
T1              T2
--------------------------------
                BEGIN
BEGIN
TryLock
COMMIT
                TryLock (fails)
```

Reply:
I think I added this schedule.